### ISSUE TITLE
name shadowing in derived types

### ISSUE THREAD

**Issue by Xeverous** [cycfi/elements#29]
Are these intended?

- `basic_button::value()` - shadows other `value()` overloads from its parent types
- `sprite::size()` shadows `image::size()`
- `view::refresh()` shadows `base_view::refresh()`

**Comment by Xeverous** [comment]
Instantiations of `indirect` cause compilation errors. `indirect` is instantiated in `hold` and similar functions.

`indirect` tries to implement wrapping interface which forwards all `element` functions but if there is any function name shadowing in derived class it causes compilation errors:

```
In file included from [...]/elements/lib/include/elements/view.hpp:16,
                 from [...]/elements/lib/include/elements.hpp:11,
                 from [...]/elements/examples/buttons/main.cpp:6:
[...]/elements/lib/include/elements/element/indirect.hpp: In instantiation of 'void cycfi::elements::indirect<Base>::value(std::__cxx11::string) [with Base = cycfi::elements::shared_element<cycfi::elements::basic_progress_bar<cycfi::elements::solid_color_element, cycfi::elements::solid_color_element, cycfi::elements::basic_progress_bar_base> >; std::__cxx11::string = std::__cxx11::basic_string<char>]':
[...]/elements/lib/include/elements/element/indirect.hpp:250:16:   required from here
[...]/elements/lib/include/elements/element/indirect.hpp:252:7: error: no matching function for call to 'cycfi::elements::basic_progress_bar<cycfi::elements::solid_color_element, cycfi::elements::solid_color_element, cycfi::elements::basic_progress_bar_base>::value(std::__cxx11::string&)'
       this->get().value(val);
       ^~~~
In file included from [...]/elements/lib/include/elements/element.hpp:22,
                 from [...]/elements/lib/include/elements.hpp:12,
                 from [...]/elements/examples/buttons/main.cpp:6:
[...]/elements/lib/include/elements/element/progress_bar.hpp:28:31: note: candidate: 'double cycfi::elements::progress_bar_base::value() const'
       double                  value() const { return _value; }
                               ^~~~~
[...]/elements/lib/include/elements/element/progress_bar.hpp:28:31: note:   candidate expects 0 arguments, 1 provided
[...]/elements/lib/include/elements/element/progress_bar.hpp:29:31: note: candidate: 'virtual void cycfi::elements::progress_bar_base::value(double)'
       void                    value(double val) override;
                               ^~~~~
[...]/elements/lib/include/elements/element/progress_bar.hpp:29:31: note:   no known conversion for argument 1 from 'std::__cxx11::string' {aka 'std::__cxx11::basic_string<char>'} to 'double'
```

So I think we need to add `using element::value;` etc in all classes which override only certain overloads.

**Comment by Xeverous** [comment]
See #49

Note that these 2 are still present, not sure what to do with them (these are not virtual):

- `sprite::size()` shadows `image::size()`
- `view::refresh()` shadows `base_view::refresh()`

**Comment by djowel** [comment]
The first two view::refresh are virtual, and they are overridden. What's the problem with that? That looks OK: `void refresh() override;` and `void refresh(rect area) override;`

**Comment by djowel** [comment]
Oddly enough, `sprite::size()` is not implemented. These should also be virtual. I'm fixing that now.

**Comment by djowel** [comment]
sprite::size is fixed in 856162b9a008253659b05a5026a95b1bea61d819

**Comment by Xeverous** [comment]
Ok, everything looks fine now. Some functions marked `virtual` could be changed to use `override` but that will be a separate issue.
